native/src/widget/text.rs

```diff
@@ -197,6 +235,158 @@ pub fn draw<Renderer>(
+    let mut get_byte_offset = |grapheme_index| {
+        byte_offsets.get(grapheme_index).copied().or_else(|| {
+            byte_offsets.extend(
+                grapheme_indices
+                    .by_ref()
+                    .take((grapheme_index - byte_offsets.len()) + 1),
+            );
+            byte_offsets.get(grapheme_index).copied()
+        })
+    };
+
+    for &Highlight { start, end, color } in highlights {
+        let start_index = if let Some(index) = get_byte_offset(start.min(end)) {
+            index
+        } else {
+            continue;
+        };
+
+        let end_index = if let Some(index) = get_byte_offset(start.max(end)) {
+            index
+        } else {
+            continue;
+        };
+
+        // The content prior to the start of the highlight is relevant for calculating offsets:
+        // The total height of all the lines of text above the line with the highlight,
+        // and the width of all the text in the line with the highlight, up until the start of the highlight.
+        let before_start = &content[..start_index];
+
+        // Iced's text layouting treats standalone carriage returns (\r) as line endings, but `str::lines`
+        // does not, so we must manually search for the relevant line ending* instead.
+        let r = before_start.rfind('\r');
+        let n = before_start.rfind('\n');
+        let before_start_linebreak = r
+            .zip(n)
+            // If `zip` returns `Some(_)`, there may be multiple line endings
+            .map(|(r, n)| match (r + 1).cmp(&n) {
+                // The rightmost line ending is `\n`
+                Ordering::Less => (n, n),
+                // The rightmost line ending is `\r\n`
+                Ordering::Equal => (r, n),
+                // The rightmost line ending is `\r`
+                Ordering::Greater => (r, r),
+            })
+            // If `zip` returns `None`, there is either 1 or 0 line endings - if 1, `xor` returns `Some(_)`
+            .or_else(|| r.xor(n).map(|i| (i, i)));
+
+        // *Get the text preceding and following the rightmost line ending
+        let (above_lines, before_start_line) = match before_start_linebreak {
+            Some((l, r)) => (&before_start[..l], &before_start[(r + 1)..]),
+            None => (Default::default(), before_start),
+        };
+
+        // Measure height of lines up until the line that contains the start of the highlight
+        let (_, mut height_offset) =
+            renderer.measure(above_lines, size, font.clone(), Size::INFINITY);
+
+        // If the highlight crosses over multiple lines, draw a seperate rect on each line
+        // BUG: This ignores single `\r` but Iced's text layouting does not (See above)
+        // BUG #2: Text wrapping caused by the text not being given wide enough bounds is not handled at all
+        //         (And furthermore it currently _can't_ be handled because there's no way to get information about it)
```

Comment:
If there is a way to handle this, please correct me, but if not, this would be a blocker on ensuring correct behaviour.